### PR TITLE
storage: bugfixes

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: lxd/storage_lvm.go [hunk 1]
```diff
@@ -759,31 +759,69 @@ func (s *storageLvm) ContainerMount(name string, path string) (bool, error) {
...
+	ourMount := false
+	if !shared.IsMountPoint(containerMntPoint) {
+		imgerr = tryMount(containerLvmPath, containerMntPoint, lvFsType, 0, mountOptions)
+		ourMount = true
+	}
+	if waitChannel, ok := lxdStorageLockMap[containerMountLockID]; ok {
+		close(waitChannel)
+		delete(lxdStorageLockMap, containerMountLockID)
+	}
+	lxdStorageLock.Unlock()
```
Comment (stgraber) [inline lxd/storage_lvm.go]: You're holding that lock for a long time here (can be up to 10s if tryMount fails)

File: lxd/storage_lvm.go [hunk 2]
```diff
@@ -759,31 +759,69 @@ func (s *storageLvm) ContainerMount(name string, path string) (bool, error) {
...
+	ourUmount := false
+	if shared.IsMountPoint(containerMntPoint) {
+		imgerr = tryUnmount(containerMntPoint, 0)
+		ourUmount = true
+	}
+	if waitChannel, ok := lxdStorageLockMap[containerUmountLockID]; ok {
+		close(waitChannel)
+		delete(lxdStorageLockMap, containerUmountLockID)
+	}
+	lxdStorageLock.Unlock()
```
Comment (stgraber) [inline lxd/storage_lvm.go]: Same comment as above

File: lxd/storage_zfs.go [hunk 3]
```diff
@@ -398,33 +398,70 @@ func (s *storageZfs) ContainerMount(name string, path string) (bool, error) {
...
+	ourMount := false
+	if !shared.IsMountPoint(containerPoolVolumeMntPoint) {
+		imgerr = s.zfsPoolVolumeMount(fs)
+		ourMount = true
+	}
+	if waitChannel, ok := lxdStorageLockMap[containerMountLockID]; ok {
+		close(waitChannel)
+		delete(lxdStorageLockMap, containerMountLockID)
+	}
+	lxdStorageLock.Unlock()
```
Comment (stgraber) [inline lxd/storage_zfs.go]: Same thing

File: lxd/storage_zfs.go [hunk 4]
```diff
@@ -398,33 +398,70 @@ func (s *storageZfs) ContainerMount(name string, path string) (bool, error) {
...
+	ourUmount := false
+	if shared.IsMountPoint(containerPoolVolumeMntPoint) {
+		imgerr = s.zfsPoolVolumeUmount(fs)
+		ourUmount = true
+	}
+	if waitChannel, ok := lxdStorageLockMap[containerUmountLockID]; ok {
+		close(waitChannel)
+		delete(lxdStorageLockMap, containerUmountLockID)
+	}
+	lxdStorageLock.Unlock()
```
Comment (stgraber) [inline lxd/storage_zfs.go]: Same thing

**Comment by brauner** [comment]
Right, I changed the lock to be released after the lock id has been added to the map. Then a new lock is taken, held until the mount succeeds and then is released.

File: lxd/images.go [hunk 5]
```diff
@@ -130,7 +130,7 @@ func unpack(d *Daemon, file string, path string, sType storageType) error {
 		// Check if we're running out of space
 		if int64(fs.Bfree) < int64(2*fs.Bsize) {
 			if sType == storageTypeLvm {
-				return fmt.Errorf("Unable to unpack image, run out of disk space (consider increasing your pool's volume.size).")
+				return fmt.Errorf("Unable to unpack image, run out of disk space.")
```
Comment (stgraber) [inline lxd/images.go]: What's the problem with the existing error message?

File: lxd/storage_btrfs.go [hunk 6]
```diff
@@ -571,23 +571,23 @@ func (s *storageBtrfs) ContainerCreateFromImage(container container, fingerprint
...
-		imageCreationInPoolLock.Unlock()
+	lxdStorageLock.Lock()
+	if waitChannel, ok := lxdStorageLockMap[imageStoragePoolLockID]; ok {
+		lxdStorageLock.Unlock()
 		if _, ok := <-waitChannel; ok {
 			shared.LogWarnf("Value transmitted over image lock semaphore?")
 		}
 	} else {
-		imageCreationInPool[imageStoragePoolLockID] = make(chan bool)
+		lxdStorageLockMap[imageStoragePoolLockID] = make(chan bool)
```
Comment (stgraber) [inline lxd/storage_btrfs.go]: Should unlock after that

File: lxd/storage_btrfs.go [hunk 7]
```diff
@@ -571,23 +571,23 @@ func (s *storageBtrfs) ContainerCreateFromImage(container container, fingerprint
...
 		var imgerr error
 		if !shared.PathExists(imageMntPoint) || !s.isBtrfsPoolVolume(imageMntPoint) {
 			imgerr = s.ImageCreate(fingerprint)
 		}
-		if waitChannel, ok := imageCreationInPool[imageStoragePoolLockID]; ok {
+		if waitChannel, ok := lxdStorageLockMap[imageStoragePoolLockID]; ok {
```
Comment (stgraber) [inline lxd/storage_btrfs.go]: And lock again before this

File: lxd/storage_lvm.go [hunk 8]
```diff
@@ -547,23 +547,23 @@ func (s *storageLvm) ContainerCreateFromImage(container container, fingerprint s
...
 			imgerr = s.ImageCreate(fingerprint)
 		}
-		if waitChannel, ok := imageCreationInPool[imageStoragePoolLockID]; ok {
+		if waitChannel, ok := lxdStorageLockMap[imageStoragePoolLockID]; ok {
 			close(waitChannel)
-			delete(imageCreationInPool, imageStoragePoolLockID)
+			delete(lxdStorageLockMap, imageStoragePoolLockID)
 		}
-		imageCreationInPoolLock.Unlock()
+		lxdStorageLock.Unlock()
```
Comment (stgraber) [inline lxd/storage_lvm.go]: Same comment as storage_btrfs wrt locking

File: lxd/storage_lvm.go (hunk 1 above)
Comment (stgraber) [inline lxd/storage_lvm.go]: This hasn't been fixed or github is lying to me again

File: lxd/images.go (hunk 5 above)
Comment (brauner) [inline lxd/images.go]: Oh that's a merge conflict on my end. You seem to have changed it. Before it was reporting a specific message refering to the old storage key.

File: lxd/storage_lvm.go (hunk 1 above)
Comment (brauner) [inline lxd/storage_lvm.go]: Nah, I pushed prematurely.

**Comment by stgraber** [comment]
jenkins: test this please

**Pull request merged**